src/tls.c

```diff
@@ -1579,134 +1579,190 @@ static int TLSX_SetALPN(TLSX** extensions, const void* data, word16 size,
+    }
+
+    if (sel != NULL) {
+        /* set the matching negotiated protocol */
+        r = TLSX_SetALPN(&ssl->extensions, sel, sel_len, ssl->heap);
+        if (r != WOLFSSL_SUCCESS) {
+            WOLFSSL_MSG("TLSX_SetALPN failed");
+            r = BUFFER_ERROR;
+            goto cleanup;
+        }
+        /* reply to ALPN extension sent from peer */
 #ifndef NO_WOLFSSL_SERVER
         TLSX_SetResponse(ssl, TLSX_APPLICATION_LAYER_PROTOCOL);
 #endif
     }
+    r = 0;
 
-    return 0;
+cleanup:
+    WOLFSSL_LEAVE("ALPN_Select", r);
+    return r;
+}
+
+/** Parses a buffer of ALPN extensions and set the first one matching
+ * client and server requirements */
+static int TLSX_ALPN_ParseAndSet(WOLFSSL *ssl, const byte *input, word16 length,
+                                 byte isRequest)
+{
+    word16  size = 0, offset = 0, wlen;
+    int     r = BUFFER_ERROR;
+    TLSX    *extension;
+    const byte *s;
+
+    if (OPAQUE16_LEN > length)
+        goto cleanup;
+
+    ato16(input, &size);
+    offset += OPAQUE16_LEN;
+
+    /* validating alpn list length */
+    if (size == 0 || length != OPAQUE16_LEN + size)
+        goto cleanup;
+    /* validating length of entries before accepting */
+    for (s = input + offset, wlen = 0; (s - input) < size; s += wlen) {
+        wlen = *s++;
+        if (wlen == 0 || (s + wlen - input) > length)
+            goto cleanup;
+    }
+
+    if (isRequest) {
+        /* keep the list sent by peer, if this is from a request. We
+         * use it later in ALPN_Select() for evaluation. */
+        if (ssl->alpn_peer_requested != NULL)
+            XFREE(ssl->alpn_peer_requested, ssl->heap, DYNAMIC_TYPE_ALPN);
+
+        ssl->alpn_peer_requested = (byte *)XMALLOC(size, ssl->heap,
+                                                   DYNAMIC_TYPE_ALPN);
+        if (ssl->alpn_peer_requested == NULL) {
+            r = MEMORY_ERROR;
+            goto cleanup;
+        }
+        ssl->alpn_peer_requested_length = size;
+        XMEMCPY(ssl->alpn_peer_requested, (char*)input + offset, size);
+    }
+    else {
+        /* a response, we should find the value in our config */
+        const byte *sel = NULL;
+        byte sel_len = 0;
+
+        r = ALPN_find_match(ssl, &extension, &sel, &sel_len, input + offset, size);
+        if (r != 0)
+            goto cleanup;
+
+        if (sel != NULL) {
+            /* set the matching negotiated protocol */
+            r = TLSX_SetALPN(&ssl->extensions, sel, sel_len, ssl->heap);
+            if (r != WOLFSSL_SUCCESS) {
+                WOLFSSL_MSG("TLSX_SetALPN failed");
+                r = BUFFER_ERROR;
+                goto cleanup;
+            }
+        }
+        /* If we had nothing configured, the response is unexpected */
+        else if (extension == NULL) {
+            r = TLSX_HandleUnsupportedExtension(ssl);
+            goto cleanup;
+        }
+        else {
+            /* have sth configured, but did not match. no error returned
```

Comment:
I don't understand this comment. What is sth?

Reply:
removed.